crates/ethrpc/src/current_block/mod.rs

```diff
@@ -16,6 +16,136 @@ use {
+    }
+
+    fn new(initial_block: BlockInfo) -> Self {
+        let (watch_sender, watch_receiver) = watch::channel(initial_block);
+        let (broadcast_sender, _broadcast_receiver) = broadcast::channel(MAX_SIZE);
+
+        Self {
+            watch_receiver,
+            watch_sender,
+            broadcast_sender,
+        }
+    }
+}
+
+/// Monitors the blockchain for new blocks and allows users to get notified
+/// about them. Can be considered the "heartbeat" of the backend.
+#[derive(Debug, Clone)]
+pub struct CurrentBlockStream(Arc<BlockStreamInner>);
+
+impl CurrentBlockStream {
+    /// Returns a reference to the current block. Keeping this reference
+    /// alive locks a mutex so it should be used as shortly as possible.
+    pub fn current(&self) -> watch::Ref<'_, BlockInfo> {
+        self.0.watch_receiver.borrow()
+    }
+
+    /// Returns a stream that always yields the most recent block (if unseen).
+    /// Also yields the current block right away and not just the next new
+    /// block. Use this if you only need to know about the current state.
+    pub fn watch_stream(&self) -> WatchStream<BlockInfo> {
+        WatchStream::new(self.0.watch_receiver.clone())
+    }
+
+    /// Returns a stream that yields every new block **AFTER** it got created.
+    /// Use this if it's required to not miss any blocks.
+    /// This stream also yields blocks that got reorged so it's not guaranteed
+    /// that the block number increases monotonically but at least it does not
+    /// go back in time since reorgs never "take away" blocks.
+    ///
+    /// Panics if it gets polled after having more than [`MAX_SIZE`] buffered
+    /// blocks.
+    pub fn buffering_stream(&self) -> impl Stream<Item = BlockInfo> {
+        let receiver = self.0.broadcast_sender.subscribe();
+        BroadcastStream::new(receiver).map(Result::unwrap)
+    }
+
+    /// Spawns a new background task that watches for new blocks and notifies
+    /// the [`CurrentBlockStream`] about them.
+    pub async fn new(url: Url, poll_interval: Duration) -> Result<Self> {
+        // Build new Web3 specifically for the current block stream to avoid batching
+        // requests together on chains with a very high block frequency.
+        let web3 = Web3::new(Web3Transport::new(HttpTransport::new(
+            Default::default(),
+            url,
+            "block_stream".into(),
+        )));
+        let web3 = instrument_with_label(&web3, "base_currentBlockStream".into());
+        let first_block = web3.current_block().await?;
+        tracing::debug!(number=%first_block.number, hash=?first_block.hash, "polled block");
+
+        let stream = CurrentBlockStream(Arc::new(BlockStreamInner::new(first_block)));
+        let stream_clone = stream.clone();
+
+        let update_future = async move {
+            let mut previous_block = first_block;
+            loop {
+                tokio::time::sleep(poll_interval).await;
```

Comment:
I probably missed or misunderstood something, but if more than 1 block was mined during that period, we might still miss the intermediate ones, doesn't it?

Reply:
Yeah, this is correct. I guess the PR description is slightly misleading. With this PR we'll add support for creating a stream that yields all blocks the stream ever saw. If the stream missed a block in the first place the subsequent streams will also not see this block. Felix suggested a good way to address this in a follow up PR though.